org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/commands/CommandBuilder.java

```diff
@@ -0,0 +1,202 @@
+        return analyzeFunction().kleisliCompose(functionFactory.createAEvalFunction(project, lang));
+    }
+
+    /**
+     * Returns a function that creates an {@link InputResult} from a String.
+     *
+     * @return the builder
+     */
+    public CommandBuilder<InputResult> input() {
+        return new CommandBuilder<>(this, description, inputFunction());
+    }
+
+    /**
+     * Returns a function that creates a {@link ParseResult} from a String.
+     *
+     * @return the builder
+     */
+    public CommandBuilder<ParseResult> parse() {
+        return new CommandBuilder<ParseResult>(this, description, parseFunction());
+    }
+
+    /**
+     * Returns a function that creates a {@link AnalyzeResult} from a String.
+     *
+     * @return the builder
+     */
+    public CommandBuilder<AnalyzeResult> analyze() {
+        return new CommandBuilder<>(this, description, analyzeFunction());
+    }
+
+    /**
+     * Returns a function that creates a parsed {@link TransformResult} from a String.
+     *
+     * @param action
+     *            the associated {@link ITransformAction}
+     * @return the builder
+     */
+    public CommandBuilder<TransformResult> transformParsed(ITransformAction action) {
+        return new CommandBuilder<>(this, description, pTransformFunction(action));
+    }
+
+    /**
+     * Returns a function that creates an analyzed {@link TransformResult} from a String.
+     *
+     * @param action
+     *            the associated {@link ITransformAction}
+     * @return the builder
+     */
+    public CommandBuilder<TransformResult> transformAnalyzed(ITransformAction action) {
+        return new CommandBuilder<>(this, description, aTransformFunction(action));
+    }
+
+    /**
+     * Returns a function that creates a parsed {@link EvaluateResult} from a String.
+     *
+     * @return the builder
+     */
+    public CommandBuilder<EvaluateResult> evalParsed() {
+        return new CommandBuilder<>(this, description, pEvaluateFunction());
+    }
+
+    /**
+     * Returns a function that creates an analyzed {@link EvaluateResult} from a String.
+     *
+     * @return the builder
+     */
+    public CommandBuilder<EvaluateResult> evalAnalyzed() {
+        return new CommandBuilder<>(this, description, aEvaluateFunction());
+    }
+
+    /**
+     * Sets the description of the created command.
+     *
+     * @param description
+     *            the description of the command
+     * @return the builder
+     */
+    public CommandBuilder<R> description(String description) {
+        return new CommandBuilder<>(this, description, function);
+    }
+
+    /**
+     * Returns an {@link IReplCommand} given a function.
+     *
+     * @return an {@link IReplCommand}
+     */
+    public IReplCommand build() {
+        return new IReplCommand() {
+            @Override
+            public IResult execute(String... arg) throws MetaborgException {
+                return function.apply(arg[0]);
```

Comment:
This limits the functionality of all commands to only one string argument. With extensibility in mind, is that going to be a problem?